src/cdk-experimental/selection/selection-set.ts

```diff
@@ -0,0 +1,123 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {TrackByFunction} from '@angular/core';
+import {Subject} from 'rxjs';
+
+interface TrackBySelection<T> {
+  isSelected(value: SelectableWithIndex<T>): boolean;
+  select(...values: Array<SelectableWithIndex<T>>): void;
+  deselect(...values: Array<SelectableWithIndex<T>>): void;
+  changed$: Subject<SelectionChange<T>>;
+}
+
+/**
+ * A selectable value with an optional index. The index is required when the selection is used with
+ * `trackBy`.
+ */
+export interface SelectableWithIndex<T> {
+  value: T;
+  index?: number;
+}
+
+/**
+ * Represents the change in the selection set.
+ */
+export interface SelectionChange<T> {
+  before: Array<SelectableWithIndex<T>>;
+  after: Array<SelectableWithIndex<T>>;
+}
+
+/**
+ * Maintains a set of selected items. Support selecting and deselecting items, and checking if a
+ * value is selected.
+ * When constructed with a `trackByFn`, all the items will be identified by applying the `trackByFn`
+ * on them. Because `trackByFn` requires the index of the item to be passed in, the `index` field is
+ * expected to be set when calling `isSelected`, `select` and `deselect`.
+ */
+export class SelectionSet<T> implements TrackBySelection<T> {
```

Comment:
Rather than introducing a new class, would it be possible to update the existing `SelectionModel` in `cdk/selection` to support what this needs?

Reply:
I can retrofit the `trackByFunction` support to `SelectionModel`, but it doesn't look very clean. Here is a draft: https://gist.github.com/yifange/faab1dcc3dfdb7497c3453bf7ceb7cef
I can polish it and fit it in if you think it is a good idea.

Reply:
Let's leave this PR as-is for now and revisit this in a follow-up